src/libstd/collections/hash/table.rs

```diff
@@ -581,6 +599,16 @@ impl<'t, K, V> FullBucket<K, V, &'t mut RawTable<K, V>> {
             v)
         }
     }
+
+    /// Remove this bucket's key and value from the hashtable.
+    /// Only used for inplacement insertion.
+    pub fn remove(&mut self) {
+        self.table.size -= 1;
+        unsafe {
+            *self.raw.hash = EMPTY_BUCKET;
+            ptr::read(self.raw.pair); // drop right now
```

Comment:
This is possibly incorrect. I think you’ll notice why if you add a test that looks like this (you probably should add one similar to the case below):

```
struct Banana<'a>(&'a mut bool);
impl Drop for Banana {
    fn drop(&mut self) {
        if !*self.0 { panic!("double drop!"); }
        *self.0 = false;
    }
}

let mut hm = HashMap::new();
let mut can_drop = true;
hm.insert(0, Banana(&mut can_drop));
hm.entry(0) <- panic!("boom") ;
// first drop happens in `make_place`, where the `Banana(true)` gets dropped and `can_drop` is set to false
// then a `*place.pointer() = panic!("boom")` is executed, which unwinds, thus dropping the place
// place destructor drops the `Banana(false)`, and thus double-panic occurs and the process aborts.
//
// In other words, current implementation of Drop reads uninitialized memory.
```